### DISCUSSION TITLE
[Idea] createNonceManager with Redis

### DISCUSSION THREAD

**Discussion by songkeys** [wevm/viem#2423] [Idea / Feature Request] [answered]
With today's [implementation](https://github.com/wevm/viem/blob/a7c1ab1d101f730e8d25bbe848304c73b43564f7/src/utils/nonceManager.ts), the `nonceManager` won't work as expected when in multiple server instances that relies on redis or an external database to share the current value of nonce.

While currently developers can implement the `createNonceManager` themselves, I think we could expose the configuration of `consume`, `increment`, `get` and `reset` methods?

Also a built-in redis interface that only requires to pass in a `redis` instance would be great as well.

**Comment by jxom** [comment, marked as answer]
Check out the `source` parameter: https://viem.sh/docs/accounts/createNonceManager#source

**Reply by songkeys** [reply]
I think the `source` only works on one server because the [in-memeory map](https://github.com/wevm/viem/blob/a7c1ab1d101f730e8d25bbe848304c73b43564f7/src/utils/nonceManager.ts#L70) is not sharable across machines?

**Reply by jxom** [reply]
The purpose of that is to manage the delta while the retrieval of the nonce is in flight (which may happen when functions are executed in the same event loop). It also reset after the source retrieval [is finalized](https://github.com/wevm/viem/blob/a7c1ab1d101f730e8d25bbe848304c73b43564f7/src/utils/nonceManager.ts#L85).
This is really to manage the case when executing transactions across an event loop on **the client**.
You’d face issues regardless of the source you use if you executed transactions on the same event loop on a server, since it would take more time to replicate the source on the external database/store. So it is recommended to execute transactions sequentially (ie. don’t use Promise.all) on the server (while using some Redis source for the nonce manager).

**Reply by songkeys** [reply]
Yes. I got you. I meant if we have multiple servers using redis to store the nonce, the current implementation won't work as expected.

For example:

- Current nonce is `1`.
- 2 requests are coming at the same time. The load balancer distributes them to server A and server B at the same time.
- These two servers retrieve the current nonce from redis, found that it's `1`.
- Both servers increments them *locally* using map to `2` and use this nonce value to call the action, which will cause transaction replacement.

I think in this case the `increment` method should call [INCR](https://redis.io/docs/latest/commands/incr/) in redis, rather than a local variable that is not able to share across machines.

**Reply by jxom** [reply]
Happy to expose `increment` on the API and see how it goes!

**Comment by DIYgod** [comment]
We are experiencing the same issue. Is there an estimated completion time or any progress for the increment API? @jxom
